File: rate-limiter/csharp/src/RateLimiter/Limiter.cs
namespace RateLimiter;

public class Limiter
{
    public const int DefaultMaxRequests = 100;
    public static readonly TimeSpan DefaultWindowDuration = TimeSpan.FromSeconds(60);

    private readonly Rule _rule;
    private readonly IClock _clock;
    private readonly Dictionary<string, WindowState> _windows = new();

    public Limiter(Rule rule, IClock clock)
    {
        _rule = rule;
        _clock = clock;
    }

    public Decision Request(string key)
    {
        var now = _clock.Now();

        if (!_windows.TryGetValue(key, out var state) || now >= state.Start + _rule.Window)
        {
            _windows[key] = new WindowState(now, 1);
            return new Decision.Allowed();
        }

        if (state.Count < _rule.MaxRequests)
        {
            _windows[key] = state with { Count = state.Count + 1 };
            return new Decision.Allowed();
        }

        var retryAfter = (state.Start + _rule.Window) - now;
        return new Decision.Rejected(retryAfter);
    }

    private record struct WindowState(DateTime Start, int Count);
}

File: rate-limiter/python/src/rate_limiter/limiter.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta

from .clock import Clock
from .decision import Allowed, Decision, Rejected
from .rule import Rule

DEFAULT_MAX_REQUESTS = 100
DEFAULT_WINDOW_DURATION = timedelta(seconds=60)


@dataclass
class _WindowState:
    start: datetime
    count: int


class Limiter:
    def __init__(self, rule: Rule, clock: Clock) -> None:
        self._rule = rule
        self._clock = clock
        self._windows: dict[str, _WindowState] = {}

    def request(self, key: str) -> Decision:
        now = self._clock.now()
        state = self._windows.get(key)

        if state is None or now >= state.start + self._rule.window:
            self._windows[key] = _WindowState(now, 1)
            return Allowed()

        if state.count < self._rule.max_requests:
            state.count += 1
            return Allowed()

        retry_after = (state.start + self._rule.window) - now
        return Rejected(retry_after)
